core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java

```diff
@@ -1202,6 +1207,83 @@ public UnifyResult apply(UnifyRuleCall call) {
     }
   }
 
+  /** Implementation of {@link UnifyRule} that matches a
+   * {@link MutableSort} on {@link MutableProject} to a
+   * {@link MutableSort} on {@link MutableProject}.
+   *
+   * <p>Example: target has same collation as that of query</p>
+   * <ul>
+   * <li>query:   Sort(sort0=[$0], dir0=[ASC])
+   *                Project(projects: [$0, $1])
+   *                  Scan(table: [hr, emps])</li>
+   * <li>target:  Sort(sort0=[$0], dir0=[ASC])
+   *                Project(projects: [$0, $1, $2])
+   *                  Scan(table: [hr, emps])</li>
+   * </ul>*/
+  private static class SortOnProjectToSortOnProjectUnifyRule extends AbstractUnifyRule {
+    public static final SortOnProjectToSortOnProjectUnifyRule INSTANCE =
+        new SortOnProjectToSortOnProjectUnifyRule();
+
+    private SortOnProjectToSortOnProjectUnifyRule() {
+      super(
+          operand(MutableSort.class,
+              operand(MutableProject.class, query(0))),
+          operand(MutableSort.class,
+              operand(MutableProject.class, target(0))), 1);
+    }
+
+    protected UnifyResult apply(UnifyRuleCall call) {
+      final MutableSort query = (MutableSort) call.query;
+      final MutableSort target = (MutableSort) call.target;
+      final MutableProject queryProject = (MutableProject) query.getInput();
+      final MutableProject targetProject = (MutableProject) target.getInput();
+      if (!queryProject.getInput().equals(targetProject.getInput())) {
```

Comment:
Instead of `equals`, we should compare the input's digest

Reply:
I checked all kinds of MutableRel nodes and found `equals` can cover the properties in `digest`.

Reply:
Thanks, i rechecked and the `equals` did work for `MutableRel`.